File: app/src/main/java/net/xzos/UpgradeAll/server/app/engine/js/JavaScriptEngine.java
package net.xzos.UpgradeAll.server.app.engine.js;

import androidx.annotation.NonNull;

import net.xzos.UpgradeAll.server.app.engine.api.EngineApi;

import org.json.JSONObject;

import java.util.ArrayList;

public class JavaScriptEngine extends EngineApi {

    private static final String TAG = "JavaScriptEngine";
    private static String[] LogObjectTag;

    private JavaScriptCoreEngine javaScriptCoreEngine;

    private int releaseNum = 0;
    private ArrayList<String> versionNumberList = new ArrayList<>();
    private ArrayList<String> changeLogList = new ArrayList<>();
    private ArrayList<JSONObject> releaseDownloadList = new ArrayList<>();

    private JavaScriptEngine(@NonNull Builder builder) {
        this.javaScriptCoreEngine = builder.javaScriptCoreEngine;
        LogObjectTag = builder.logObjectTag;
        if (builder.enableLogJsCode) {
            Log.i(LogObjectTag, TAG, String.format("JavaScriptCoreEngine: jsCode: \n%s", builder.jsCode));  // 只打印一次 JS 脚本
        }
    }

    public static class Builder {
        private final JavaScriptCoreEngine javaScriptCoreEngine;
        private final String[] logObjectTag;
        private final String jsCode;
        private boolean enableLogJsCode;

        public Builder(String[] logObjectTag, String URL, String jsCode) {
            this.javaScriptCoreEngine = new JavaScriptCoreEngine(logObjectTag, URL, jsCode);
            this.logObjectTag = javaScriptCoreEngine.getLogObjectTag();
            this.jsCode = jsCode;
        }

        public Builder enableLogJsCode(boolean enableLogJsCode) {
            this.enableLogJsCode = enableLogJsCode;
            return this;
        }

        public JavaScriptEngine build() {
            return new JavaScriptEngine(this);
        }
    }

    @Override
    public void refreshData() {
        releaseNum = getReleaseNum();
        for (int i = 0; i < releaseNum; i++) {
            versionNumberList.add(getVersioning(i));
            changeLogList.add(getChangelog(i));
            releaseDownloadList.add(getReleaseDownload(i));
        }
    }

    @Override
    public int getReleaseNum() {
        if (this.releaseNum == 0) {
            try {
                this.releaseNum = javaScriptCoreEngine.getReleaseNum();
            } catch (Throwable e) {
                Log.e(LogObjectTag, TAG, "getReleaseNum: 脚本执行错误, ERROR_MESSAGE: " + e.toString());
            }
        }
        return this.releaseNum;
    }

    @Override
    public String getVersioning(int releaseNum) {
        String versionNumber = null;
        if (versionNumberList.size() == 0) {
            try {
                versionNumber = javaScriptCoreEngine.getVersioning(releaseNum);
            } catch (Throwable e) {
                Log.e(LogObjectTag, TAG, "getVersioning: 脚本执行错误, ERROR_MESSAGE: " + e.toString());
                return null;
            }
        } else if (releaseNum >= 0 && releaseNum < versionNumberList.size())
            versionNumber = versionNumberList.get(releaseNum);
        return versionNumber;
    }

    @Override
    public String getChangelog(int releaseNum) {
        String changeLog = null;
        if (changeLogList.size() == 0) {
            try {
                changeLog = javaScriptCoreEngine.getChangelog(releaseNum);
            } catch (Throwable e) {
                Log.e(LogObjectTag, TAG, "getChangelog: 脚本执行错误, ERROR_MESSAGE: " + e.toString());
                return null;
            }
        } else if (releaseNum >= 0 && releaseNum < changeLogList.size())
            changeLog = changeLogList.get(releaseNum);
        return changeLog;
    }

    @Override
    public JSONObject getReleaseDownload(int releaseNum) {
        JSONObject releaseDownload = null;
        if (releaseDownloadList.size() == 0) {
            try {
                releaseDownload = javaScriptCoreEngine.getReleaseDownload(releaseNum);
            } catch (Throwable e) {
                Log.e(LogObjectTag, TAG, "getReleaseDownload: 脚本执行错误, ERROR_MESSAGE: " + e.toString());
                return null;
            }
        } else if (releaseNum >= 0 && releaseNum < releaseDownloadList.size())
            releaseDownload = releaseDownloadList.get(releaseNum);
        return releaseDownload;
    }

    @Override
    public String getDefaultName() {
        try {
            return javaScriptCoreEngine.getDefaultName();
        } catch (Throwable e) {
            Log.e(LogObjectTag, TAG, "getDefaultName: 脚本执行错误, ERROR_MESSAGE: " + e.toString());
        }
        return null;
    }
}

File: app/src/main/kotlin/net/xzos/upgradeAll/server/app/engine/js/JavaScriptEngine.kt
package net.xzos.upgradeAll.server.app.engine.js

import net.xzos.upgradeAll.data.json.gson.JSReturnData
import net.xzos.upgradeAll.data.json.nongson.JSCache
import net.xzos.upgradeAll.data.json.nongson.ObjectTag
import net.xzos.upgradeAll.server.log.LogUtil

class JavaScriptEngine internal constructor(
        private val objectTag: ObjectTag,
        URL: String?,
        jsCode: String?,
        debugMode: Boolean = false
) {

    private val jsCache = JSCache(objectTag)
    private val javaScriptCoreEngine = JavaScriptCoreEngine(objectTag, URL, jsCode, jsCache)

    init {
        if (!debugMode) {
            Log.i(objectTag, TAG, "JavaScriptCoreEngine: jsCode: \n$jsCode")  // 只打印一次 JS 脚本
        }
        javaScriptCoreEngine.jsUtils.debugMode = debugMode

        jsCache.clearCache()  // 初始化类时，清除旧的缓存
    }

    suspend fun getDefaultName(): String? = javaScriptCoreEngine.getDefaultName()

    suspend fun getAppIconUrl(): String? = javaScriptCoreEngine.getAppIconUrl()

    /**
     * 返回包含版本信息的数据类列表
     */
    suspend fun getJsReturnData(): JSReturnData =
            jsCache.getJsReturnData()
                    ?: javaScriptCoreEngine.getReleaseInfo().also {
                        jsCache.cacheJsReturnData(it)
                    }

    suspend fun downloadReleaseFile(downloadIndex: Pair<Int, Int>): Boolean {
        val jsSuccessDownload = javaScriptCoreEngine.downloadReleaseFile(downloadIndex)
        return if (jsSuccessDownload)
            true
        else
            downloadFileByReleaseInfo(downloadIndex)
    }

    suspend fun downloadFileByReleaseInfo(downloadIndex: Pair<Int, Int>, externalDownloader: Boolean = false): Boolean {
        Log.e(objectTag, TAG, "downloadFile: 尝试直接下载")
        val assets = getJsReturnData().releaseInfoList[downloadIndex.first].assets
        val fileIndex = downloadIndex.second
        val fileName = assets[fileIndex].name
        val downloadUrl = assets[fileIndex].download_url
        javaScriptCoreEngine.jsUtils.downloadFile(fileName, downloadUrl, externalDownloader = externalDownloader)
        return true
    }

    companion object {
        private const val TAG = "JavaScriptEngine"
        private val Log = LogUtil
    }
}
